boards/x86: add ACRN UOS sample board configuration (#14530)
zephyrproject-rtos/zephyr: boards/x86/acrn/doc/index.rst

.. _acrn:

> ACRN UOS (Guest Operating System)

Reviewer: Nitpick: I think it's preferable to use `User` here so it matches the `UOS` acronym we use throughout this doc.
```suggestion
ACRN UOS (User Operating System)
```

Author: done

running Zephyr RTOS applications.

ACRN is a flexible, lightweight reference hypervisor, built with real-time
and safety-criticality in mind, optimized to streamline embedded development
through an open source platform. Check out the `Introduction to Project ACRN
<https://projectacrn.github.io/latest/introduction/>`_ for more information.

This baseline configuration can be used as a starting point for creating
custom ACRN UOS configurations. It currently supports the following devices:

* I/O APIC
* local APIC timer
* NS16550 UARTs

Serial Ports
------------

The serial ports are assumed present at traditional ``COM1:`` and ``COM2:``
I/O-space addresses (based at ``0x3f8`` and ``0x2f8``, respectively). Only
polled operation is supported in this baseline configuration, as IRQ
assignments under ACRN are configurable (and frequently non-standard).
Interrupt-driven and MMIO operation are also possible.

Building and Running
********************

#. The build process for ACRN targets is similar to other boards.  For example,
you can build the :ref:`hello_world` sample for ACRN with:

.. zephyr-app-commands::
   :zephyr-app: samples/hello_world
   :board: acrn

>    :goals: build

Reviewer: Assuming we can also use `west` to build this, we should add this option `:tool: all`

Reviewer: > Assuming we can also use `west` to build this, we should add this option `:tool: all`

I have just tested building the `hello_world` sample using `west` and it worked, let's add `:tool: all` to this section.

Author: done

#. Build grub2 boot loader image

>    We can build grub2 bootloader for zephyr using

Reviewer: ```suggestion
   We can build grub2 bootloader for Zephyr using
```

Author: done

   ``boards/x86/common/scripts/build_grub.sh``
   which locate in `Zephyr Sourcecode
   <https://github.com/zephyrproject-rtos/zephyr>`_.

   .. code-block:: none

      $ ./boards/x86/common/scripts/build_grub.sh x86_64

>    Find the binary at boards/x86/common/scripts/grub/bin/grub_x86_64.efi.

Reviewer: ```suggestion
   Find the binary at ``boards/x86/common/scripts/grub/bin/grub_x86_64.efi``
```

Author: done

#. Preparing the boot device

   .. code-block:: none

>       $ dd if=/dev/zero of=zephyr.img bs=1M count=1024

Reviewer: This creates a 1GB image, and we don't even need 1.5M with all the files. Let's create a smaller one:
```suggestion
      $ dd if=/dev/zero of=zephyr.img bs=1M count=35
```
**Note**: I am using 35MB on purpose here as this is the smallest partition size allowed by `FAT32`.

Author: done

      $ mkfs.vfat -F 32 zephyr.img

>       $ losetup -f -P --show zephyr.img

Reviewer: We need `sudo` here, and also need to account for the fact that the loop device may not be `/dev/loop0` in the next step.

```suggestion
      $ LOOP_DEV=`sudo losetup -f -P --show zephyr.img`
```

Author: changed

>       $ mount /dev/loop0 /mnt

Reviewer: ```suggestion
      $ sudo mount $LOOP_DEV /mnt
```

Author: changed.

>    Create the following directories.

Reviewer 2: I'd rather see the actual commands that are used (use a .. code-block) to do these actions instead of a description of what needs to be done.

```
.. code-block:: none

  $ mkdir -p /mnt/efi/boot
  $ cp grub_x86_64.efi /mnt/efi/boot

  $ mkdir -p /mnt/kernel 
  $ cp zephyr.elf /mnt/kernel

  ... etc.
```

Reviewer: Actually, better to use this:
```
sudo cp boards/x86/common/scripts/grub/bin/grub_x86_64.efi /mnt/efi/boot/bootx64.efi
```
Using `bootx64.efi` as the target name will make it so that the `OVMF.fd` firmware will automatically find and load it (so it saves us typing stuff on the EFI shell)

Author: done

   :file:`/mnt/efi`

   :file:`/mnt/efi/boot`

   :file:`/mnt/kernel/`

   Copy ``zephyr.elf`` to ``/mnt/kernel/`` and ``grub_x86_64.efi`` to
   ``/mnt/efi/boot/``.

   Create ``/mnt/efi/boot/grub.cfg`` containing the following:

   .. code-block:: console

     set default=0
     set timeout=10

     menuentry "Zephyr Kernel" {
         multiboot /kernel/zephyr.elf
     }

   .. note::

>       ``zephyr.elf`` locates in the ``build/zephyr/``. For example, we can find

Reviewer 2: ```suggestion
      ``zephyr.elf`` is found in the sample's ``build/zephyr/`` folder. For example, we can find
```

Author: done

      ``zephyr.elf`` in ``samples/hello_world/build/zephyr/zephyr.elf`` since
      we built the ``Hello World App``.

#. Follow `Getting started guide for Intel NUC
   <https://projectacrn.github.io/latest/getting-started/apl-nuc.html>`_ to
   boot "The ACRN Service OS"

#. Boot Zephyr as User OS

>    Make one new directory ``zephyr`` in SOS. And then copy zephyr.img,

Author: changed

>    Make one new directory ``zephyr`` in SOS. And then copy ``zephyr.img``,

Reviewer 2: Same comment here. Don't describe the actions needed, show the actual commands in a .. code-block

   ``/usr/share/acrn/samples/nuc/launch_zephyr.sh`` and
   ``/usr/share/acrn/samples/bios/`` to ``zephyr`` directory. We can find
   ``OVMF.fd`` and ``launch_zephyr.sh`` in `ACRN Repo
   <https://github.com/projectacrn/acrn-hypervisor>`_ also.

   Run the ``launch_zephyr.sh`` script to launch Zephyr as UOS.

   .. code-block:: none

>       $ ./launch_zephyr.sh

Reviewer: Need to be `root`:
```suggestion
      $ sudo ./launch_zephyr.sh
```

Author: done

>    Load Zephyr using ``grub_x86_64.efi`` in UEFI shell:

Reviewer: You can get rid of this line (and the following code-block) if you use `bootx64.efi` as the target name (see above)

Author: done

   .. code-block:: console

      $ shell> FS0:
      $ FS0:\> efi\boot\grub_x86_64.efi

>    Then zephyr will boot up automatically. You will see the prompt as

Reviewer 2: ```suggestion
   Then Zephyr will boot up automatically. You will see the prompt as
```

Author: changed

      $ sudo mkdir -p /mnt/kernel
      $ sudo cp samples/hello_world/build/zephyr/zephyr.elf /mnt/kernel

   Create ``/mnt/efi/boot/grub.cfg`` containing the following:

   .. code-block:: console

     set default=0
     set timeout=10

     menuentry "Zephyr Kernel" {
         multiboot /kernel/zephyr.elf
     }

   And then unmount the image file:

   .. code-block:: console

      $ sudo umount /mnt

   You now have a virtual disk image with a bootable Zephyr in ``zephyr.img``.
   If the Zephyr build system is not the ACRN SOS, then you will need to
   transfer this image to the ACRN SOS (via, e.g., a USB stick or network).

On the ACRN SOS
---------------

#. If you are not already using the ACRN SOS, follow `Getting started guide
   for Intel NUC
   <https://projectacrn.github.io/latest/getting-started/apl-nuc.html>`_ to
   install and boot "The ACRN Service OS".

#. Boot Zephyr as User OS

   On the ACRN SOS, prepare a directory and populate it with Zephyr files.

   .. code-block:: none

      $ mkdir zephyr
      $ cd zephyr
      $ cp /usr/share/acrn/samples/nuc/launch_zephyr.sh .

>       $ cp -r /usr/share/acrn/samples/bios .

Reviewer: The bios is in `/usr/share/acrn/bios/` (and copying the entire directory means the launch script won't actually find the bios because it has to be in the same folder, not under `./bios/`, so make this line:
```suggestion
      $ cp -r /usr/share/acrn/bios/OVMF.fd .
```
instead and it should work.

Author: done. left out the -r as well since target is a file.

   You will also need to copy the ``zephyr.img`` created in the first
   section into this directory. Then run ``launch_zephyr.sh`` script
   to launch the Zephyr as a UOS.

   .. code-block:: none

      $ sudo ./launch_zephyr.sh

   Then Zephyr will boot up automatically. You will see the banner:

   .. code-block:: console

>       Hello World! ACRN

Reviewer: ```suggestion
      Hello World! acrn
```

Author: done

.. _acrn:

ACRN UOS (Guest Operating System)
#################################

Overview
********

> This baseline configuration is intended to be used as a starting point for

Reviewer 2: This could use more introduction since project ACRN is likely an unknown thing to most Zephyr developers.

Something like this perhaps:
```
This board configuration defines an ACRN User OS execution environment for
running Zephyr RTOS applications.

ACRN is a flexible, lightweight reference hypervisor, built with real-time
and safety-criticality in mind, optimized to streamline embedded
development through an open source platform. Check out the
`Introduction to Project ACRN <https://projectacrn.github.io/latest/introduction/>`_
for more information.

This baseline configuration can be used as a starting point for creating
custom ACRN UOS configurations.  It currently supports the following devices: 
```

custom ACRN UOS configurations. It supports the following devices:

* I/O APIC
* local APIC timer
* NS16550 UARTs

Serial Ports
------------

> The serial ports are assumed present at traditional COM1: and COM2: I/O-space

Reviewer 2: ```suggestion
The serial ports are assumed present at traditional ``COM1:`` and ``COM2:`` I/O-space
```

addresses (based at 0x3f8 and 0x2f8, respectively). Only polled operation is
supported in this baseline configuration, as IRQ assignments under ACRN are
configurable (and frequently non-standard). Interrupt-driven and MMIO operation

> are both theoretically possible, however.

Reviewer 2: ```suggestion
are also possible.
```

> Running

Reviewer 2: ```suggestion
Building and Running
```
(and adjust the title underline)

*******

> The output from the build process is a zephyr.elf file, which is loaded

Reviewer 2: ```suggestion
The output from the build process is a ``zephyr.elf file``, which is loaded
```

> The output from the build process is a zephyr.elf file, which is loaded

Reviewer 2: Add an example of how to build one of the existing sample applications (e.g., hello world), or do you have some other UOS sample that's more ACRN specific you can supply?

.. _acrn:

ACRN UOS (Guest Operating System)
#################################

Overview
********

This board configuration defines an ACRN User OS execution environment for
running Zephyr RTOS applications.

ACRN is a flexible, lightweight reference hypervisor, built with real-time
and safety-criticality in mind, optimized to streamline embedded development
through an open source platform. Check out the `Introduction to Project ACRN
<https://projectacrn.github.io/latest/introduction/>`_ for more information.

This baseline configuration can be used as a starting point for creating
custom ACRN UOS configurations. It currently supports the following devices:

* I/O APIC
* local APIC timer
* NS16550 UARTs

Serial Ports
------------

The serial ports are assumed present at traditional ``COM1:`` and ``COM2:``
I/O-space addresses (based at ``0x3f8`` and ``0x2f8``, respectively). Only
polled operation is supported in this baseline configuration, as IRQ
assignments under ACRN are configurable (and frequently non-standard).
Interrupt-driven and MMIO operation are also possible.

Building and Running
********************

> The build process for ACRN targets is standard.  For example::

Author: Updated as requested.

    cd $ZEPHYR_BASE/samples/hello_world
    mkdir build
    cd build
    cmake -DBOARD=acrn ..
    make

The output from the build process is a ``zephyr.elf`` file, which is loaded
by ACRN through EFI/GRUB on a virtual disk. The ACRN project has detailed
instructions for `Using Zephyr as User OS

> <https://github.com/projectacrn/acrn-hypervisor/blob/5071cd7efa3614a2c5ff94994c986e9b14c92e9b/doc/tutorials/using_zephyr_as_uos.rst>`_.

Author: Would you suggest that I put the link in (broken, and mention that it's coming soon), or simply omit it for the time being?

> <https://github.com/projectacrn/acrn-hypervisor/blob/5071cd7efa3614a2c5ff94994c986e9b14c92e9b/doc/tutorials/using_zephyr_as_uos.rst>`_.

Author: I've omitted it for the time being.  Once the ACRN doc is merged/published I'll update our docs to refer to it.
